### PR TITLE
qt{5,6}: update for the new Darwin SDK

### PR DESCRIPTION
**Pull request by reckenrode**
Qt 5.15 needs to know the exact location of the OpenGL framework, so it needs updated for the new SDK pattern. I also updated Qt 6. Both are examples of how the new Darwin SDK can simplify derivations and improve support (e.g., Qt WebEngine builds again on Darwin).

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by reckenrode** [comment]
@K900 pinging you since you’d been interested in this on Matrix.

**Comment by emilazy** [comment]
For future cleaner‐uppers, to avoid making Randy build Qt again: `5.15/qtwebengine-darwin-checks.patch` can probably go now, and `5.15/qtwebengine-darwin-no-platform-check.patch` can probably be replaced by just rewriting the hardcoded `/usr/bin/xcodebuild` path to point to our `xcbuild`.

**Review by K900 — APPROVED** [review]
This is very cool.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix [hunk 1]
```diff
@@ -257,40 +238,10 @@ qtModule ({
...
-    SecurityInterface
-    Vision
-    CoreML
-    OpenDirectory
-    Accelerate
-
-    openbsm
-    libunwind
-  ];
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [ cctools.libtool ];
```
Comment (emilazy) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:241]: This comment seems probably inaccurate now? Though I don’t really know what it meant to begin with.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix [hunk 2]
```diff
@@ -257,40 +238,10 @@ qtModule ({
...
 
   buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [
     cups
```
Comment (emilazy) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:244]: (I wonder if it just needs the headers?)

**Review by emilazy — COMMENTED** [review]
Reiterating what I said on Matrix: these kinds of diffs are incredibly satisfying and I can’t think of anything better to point people at to explain how much of an improvement the new model is. A couple comments that are more me thinking out loud than anything.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix (hunk 1 above)
Comment (reckenrode) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:241]: I’m seeing if I can drop it now that xcbuild includes `xcrun`. I also tried dropping the patches as suggested above.

**Comment by emilazy** [comment]
`0002-qtbase-qmake-fix-mkspecs-for-darwin.patch` also looks a lot like something that could be dropped or be made much smaller now that we have an actual SDK.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix (hunk 1 above)
Comment (emilazy) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:241]: Ah, I meant this comment that I don’t really understand:
```nix
  # FIXME These dependencies shouldn't be needed but can't find a way
  # around it. Chromium pulls this in while bootstrapping GN.
```

Comment (reckenrode) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:241]: I’m seeing if I can go one step further and delete the whole thing. If not, I’ll drop or change the comment.

Comment (reckenrode) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:241]: Nope. Deleting those would be non-trivial. It requires replacing the qtbase mkspecs patches with different patches to use xcbuild. If someone wants to do that later, that’s fine; but it’s beyond the scope of what I want to do with this.

File: pkgs/development/libraries/qt-5/modules/qtbase.nix [hunk 3]
```diff
@@ -141,15 +146,13 @@ stdenv.mkDerivation (finalAttrs: ({
...
-            -e 's|/usr/bin/xcode-select|xcode-select|' \
-            -e 's|/usr/bin/xcrun|xcrun|' \
-            -e 's|/usr/bin/xcodebuild|xcodebuild|' \
-            -e 's|QMAKE_CONF_COMPILER=`getXQMakeConf QMAKE_CXX`|QMAKE_CXX="clang++"\nQMAKE_CONF_COMPILER="clang++"|' \
-            ./configure
-            substituteInPlace ./mkspecs/common/mac.conf \
-                --replace "/System/Library/Frameworks/OpenGL.framework/" "${OpenGL}/Library/Frameworks/OpenGL.framework/" \
-                --replace "/System/Library/Frameworks/AGL.framework/" "${AGL}/Library/Frameworks/AGL.framework/"
+      substituteInPlace configure \
+        --replace-fail '/usr/bin/xcode-select' '${lib.getBin xcbuild}/bin/xcode-select' \
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/modules/qtbase.nix:150]: We could also use lib.getExe'

**Review by SuperSandro2000 — COMMENTED** [review]
LGTM from my limited Darwin knowledge

**Review by emilazy — APPROVED** [review]
LGTM, but I’ll let @K900 do the merge.

**Pull request merged**